iml-warp-drive/src/main.rs

```diff
@@ -0,0 +1,198 @@
+
+            let mut locks = lock_state.lock().unwrap();
+            let locks = locks.deref_mut();
+
+            locks::update_locks(locks, lock_change);
+
+            log::debug!("Current locks {:?}", locks);
+
+            send_message(serde_json::to_string(&locks).unwrap(), &user_state.clone());
+
+            Ok(())
+          })
+          .map_err(failure::Error::from)
+      })
+      .map_err(|err| eprintln!("An error occured: {}", err)),
+    );
+
+    let state = warp::any().map(move || user_state2.clone());
+
+    // GET /locks -> messages stream
+    let messaging = warp::get2()
+      .and(warp::sse())
+      .and(state)
+      .map(|sse: warp::sse::Sse, state| {
+        // reply using server-sent events
+        let stream = user_connected(state);
+        sse.reply(warp::sse::keep(stream, None))
+      });
+
+    let routes = messaging;
+
+    log::info!("about to serve");
+
+    warp::serve(routes).bind(([127, 0, 0, 1], env::get_server_port()))
+  }));
+}
+
+fn user_connected(
+  state: SharedUsers,
+) -> impl Stream<Item = impl ServerSentEvent, Error = warp::Error> {
+  // Use a counter to assign a new unique ID for this user.
+  let my_id = NEXT_USER_ID.fetch_add(1, Ordering::Relaxed);
+
+  log::debug!("User connected {}", my_id);
+
+  // Use an unbounded channel to handle buffering and flushing of messages
+  // to the event source...
+  let (tx, rx) = mpsc::unbounded();
+
+  match tx.unbounded_send(Message::UserId(my_id)) {
+    Ok(()) => (),
+    Err(_disconnected) => {
+      // The tx is disconnected, our `user_disconnected` code
+      // should be happening in another task, nothing more to
+      // do here.
+    }
+  }
+
+  // Make an extra clone of users list to give to our disconnection handler...
+  let state2 = state.clone();
+
+  // Save the sender in our list of connected users.
+  state2.lock().unwrap().insert(my_id, tx);
+
+  // Create channel to track disconnecting the receiver side of events.
+  // This is little bit tricky.
+  let (mut dtx, mut drx) = oneshot::channel::<()>();
+
+  // When `drx` will dropped then `dtx` will be canceled.
+  // We can track it to make sure when the user leaves chat.
+  warp::spawn(poll_fn(move || dtx.poll_cancel()).map(move |_| {
+    user_disconnected(my_id, &state2);
+  }));
+
+  // Convert messages into Server-Sent Events and return resulting stream.
+  rx.map(|msg| match msg {
+    Message::UserId(my_id) => (warp::sse::event("user"), warp::sse::data(my_id)).into_a(),
+    Message::Data(reply) => warp::sse::data(reply).into_b(),
+  })
+  .map_err(move |_| {
+    // Keep `drx` alive until `rx` will be closed
+    drx.close();
+    unreachable!("unbounded rx never errors");
+  })
+}
+
+fn send_message(msg: String, state: &SharedUsers) {
+  log::debug!("Sending message {:?} to users {:?}", msg, state);
+
+  for (_, tx) in state.lock().unwrap().iter() {
+    match tx.unbounded_send(Message::Data(msg.clone())) {
+      Ok(()) => (),
```

Comment:
What does this do?

Reply:
Basically consumes the result and ignores it.